File: leetcode/2_add-two-numbers/simple.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	start := &ListNode{}
	node := start
	for {
		l1Next := (*ListNode)(nil)
		l2Next := (*ListNode)(nil)
		if l1 != nil {
			node.Val += l1.Val
			l1Next = l1.Next
		}
		if l2 != nil {
			node.Val += l2.Val
			l2Next = l2.Next
		}
		if l1Next == nil && l2Next == nil && node.Val < 10 {
			break
		}
		node.Next = &ListNode{Val: node.Val / 10}
		node.Val %= 10
		node = node.Next
		l1 = l1Next
		l2 = l2Next
	}
	return start
}

File: leetcode/19_remove-nth-node-from-end-of-list/simple.cpp
// https://leetcode.com/problems/remove-nth-node-from-end-of-list/submissions/984594120/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* lastNode = head;
        while (n > 0) {
            lastNode = lastNode->next;
            n--;
        }
        if (lastNode == nullptr) {
            return head->next;
        }
        lastNode = lastNode->next;
        ListNode* beforeNthNodeFromEnd = head;
        while (lastNode) {
            lastNode = lastNode->next;
            beforeNthNodeFromEnd = beforeNthNodeFromEnd->next;
        }
        beforeNthNodeFromEnd->next = beforeNthNodeFromEnd->next->next;
        return head;
    }
};
